j5/components/gpio_pin.py

```diff
@@ -0,0 +1,191 @@
+        self.mode = initial_mode
+
+    @staticmethod
+    def interface_class() -> Type[GPIOPinInterface]:
+        """Get the interface class that is required to use this component."""
+        return GPIOPinInterface
+
+    def _require_pin_modes(self, pin_modes: List[GPIOPinMode]) -> None:
+        """Ensure that this pin is in the specified hardware mode."""
+        if not any(self.mode == mode for mode in pin_modes) and not len(pin_modes) == 0:
+            raise BadGPIOPinMode(
+                f"Pin {self._identifier} needs to be in one of {pin_modes}",
+            )
+
+    @property
+    def mode(self) -> GPIOPinMode:
+        """Get the hardware mode of this pin."""
+        return self._backend.get_gpio_pin_mode(self._board, self._identifier)
+
+    @mode.setter
+    def mode(self, pin_mode: GPIOPinMode) -> None:
+        """Set the hardware mode of this pin."""
+        if pin_mode not in self._supported_modes:
+            raise NotSupportedByHardware(
+                f"Pin {self._identifier} on {str(self._board)} \
+                does not support {str(pin_mode)}.",
+            )
+        self._backend.set_gpio_pin_mode(self._board, self._identifier, pin_mode)
+
+    @property
+    def digital_state(self) -> bool:
+        """Get the digital state of the pin."""
+        self._require_pin_modes([
+            GPIOPinMode.DIGITAL_OUTPUT,
+            GPIOPinMode.DIGITAL_INPUT,
+            GPIOPinMode.DIGITAL_INPUT_PULLUP,
+            GPIOPinMode.DIGITAL_INPUT_PULLDOWN],
+        )
+
+        # Behave differently depending on the hardware mode.
+        if self.mode is GPIOPinMode.DIGITAL_OUTPUT:
+            return self._backend.get_gpio_pin_digital_state(self._board, self._identifier)
+
+        return self._backend.read_gpio_pin_digital_state(self._board, self._identifier)
+
+    @digital_state.setter
+    def digital_state(self, state: bool) -> None:
+        """Set the digital state of the pin."""
+        self._require_pin_modes([GPIOPinMode.DIGITAL_OUTPUT])
+        self._backend.write_gpio_pin_digital_state(self._board, self._identifier, state)
+
+    @property
+    def analogue_value(self) -> float:
+        """Get the scaled analogue reading of the pin."""
+        self._require_pin_modes([GPIOPinMode.ANALOGUE_INPUT])
+        return self._backend.read_gpio_pin_analogue_value(self._board, self._identifier)
+
+    @analogue_value.setter
+    def analogue_value(self, new_value: float) -> None:
+        """Set the analogue value of the pin."""
+        self._require_pin_modes([
+            GPIOPinMode.ANALOGUE_OUTPUT,
+            GPIOPinMode.PWM_OUTPUT,
+        ])
+        if new_value < 0 or new_value > 1:
+            raise ValueError("An analogue pin value must be between 0 and 1.")
+
+        if self.mode is GPIOPinMode.ANALOGUE_OUTPUT:
+            self._backend.write_gpio_pin_dac_value(
+                self._board,
+                self._identifier,
+                new_value,
+            )
```

Comment:
Think you need a `return` at the end of this if check, otherwise we'll still execute the PWM specific code.

Reply:
Why not simply use an `else` block here?

Reply:
👍 Or that, either works. `else` is probably more explicit for the next dev who finds it.